cranelift/isle/isle/src/trie.rs

```diff
@@ -335,7 +374,13 @@ impl<'a> TermFunctionsBuilder<'a> {
     }
 
     fn build(&mut self) {
-        for rule in 0..self.termenv.rules.len() {
+        // Sort rules by priority, descending, and insert in that order.
+        let mut rule_indices: Vec<usize> = (0..self.termenv.rules.len()).collect();
+        // Sort by *negative* priority so we get descending order
+        // (highest priority first).
+        rule_indices.sort_by_key(|&index| -self.termenv.rules[index].prio.unwrap_or(0));
```

Comment:
I haven't looked at the surrounding context, but I'd guess that this expression is a single memory access and maybe a couple of arithmetic instructions. So I'd guess that copying the cached keys into another array wouldn't save any time and would waste some space.

If we're micro-optimizing I'd go the other direction and use `sort_unstable_by_key`. If you want a stable sort, in this case you can get it by making the key be `(prio, index)`.

Unrelated: you might consider using `std::cmp::Reverse` instead of negating the priority. IMO that serves as better documentation, enough that you could maybe delete the comment. I also wouldn't be surprised if it generates better code (by swapping the order of comparison rather than doing any negation) and it avoids worries about integer overflow for negating the most-negative signed value.

Reply:
Oh, neat, I didn't know about `std::cmp::Reverse` -- thanks! Updated to use that. (And agreed that caching likely doesn't make much sense here: it would just be avoiding the `unwrap_or(0)` but that's something like 3 instructions (cmp, generate 0, conditional-move) at the cost of more allocation and memory traffic.